### PR TITLE
implement sign+verify tests and test vectors

### LINKED ISSUE #5: implement verify algorithm
see https://eprint.iacr.org/2017/965.pdf section 2.2

depends on https://github.com/ChainSafe/W-OTS/issues/2, #3

### LINKED ISSUE #9: implement test vectors
see https://github.com/xx-labs/sleeve/blob/main/wots/key_test.go#L269

### LINKED ISSUE #11: Make fields of `Keys` public or add methods for them
There are some traits of substrate require to get the pk/sk of our keypair.

https://github.com/ChainSafe/xx-primitives/issues/7

For layouts of signature scheme in rust, can take a look at https://github.com/dalek-cryptography/ed25519-dalek

### PR DESCRIPTION
**Pull request by noot**
## Changes

- implement `params.verify`
- add unit tests for signing then verifying, both with and without calling `key.generate()`
- add test vectors from https://github.com/xx-labs/sleeve/blob/64e92defa018b5000d8224c2340acec06310f493/wots/key_test.go#L272
- add `ComputeLaddersMode` for `compute_ladders` instead of having bools

## Tests

```
cargo test
```

## Issues

- closes #5 
- closes #9
- closes #11

### CODE REVIEW

File: src/params.rs [hunk 1]
```diff
@@ -213,19 +254,18 @@ impl<PRFH: Hasher, MSGH: Hasher> Params<PRFH, MSGH> {
...
     fn compute_chain(
-        &mut self,
+        &self,
         p_seed: &[u8],
         input: &[u8],
         random_elements: &[Vec<u8>],
+        mut maybe_chains: Option<&mut [Vec<u8>]>,
+        idx: usize,
         begin: u8,
         end: u8,
```
Comment (clearloop) [inline src/params.rs:264]: mb we can use https://doc.rust-lang.org/std/ops/struct.Range.html instead, to save this `this function has too many arguments (8/7)`, or `range: (u8, u8)`

File: src/params.rs [hunk 2]
```diff
@@ -237,26 +277,55 @@ impl<PRFH: Hasher, MSGH: Hasher> Params<PRFH, MSGH> {
...
             hasher.sum(&mut buf);
-            curr_value.clone_from_slice(&buf);
+            curr_value.clone_from_slice(&buf[0..self.n]);
 
-            if generate {
-                chains[j as usize].copy_from_slice(&curr_value);
+            if let Some(ref mut chains) = maybe_chains {
+                let from = idx * self.n;
+                let to = (idx + 1) * self.n;
+                chains[j as usize + 1][from..to].copy_from_slice(&curr_value);
```
Comment (clearloop) [inline src/params.rs:287]: we may need to add some length check before this kind of slice operations? for example

```rust
if chains.len() > end || chains.len() < start {
  return Err(OutOfRange)
}
```

File: src/keys.rs [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
 pub const PK_SIZE: usize = 32;
 
-pub struct Key<PRFH: Hasher, MSGH: Hasher> {
-    p_seed: [u8; SEED_SIZE],
-    chains: Option<Vec<Vec<u8>>>,
+pub struct Key<PRFH: Hasher + Clone, MSGH: Hasher + Clone> {
+    pub p_seed: [u8; SEED_SIZE],
+    pub chains: Option<Vec<Vec<u8>>>,
     secret_key: Vec<u8>,
     public_key: Option<Vec<u8>>,
```
Comment (clearloop) [inline src/keys.rs:14]: Is it possible to make the `public_key` and `secret_key` `pub` also, see https://github.com/dalek-cryptography/ed25519-dalek/blob/ad461f4f0f56507a2945510dedb914b3d3547c40/src/keypair.rs#L36

really need to get these two keys from `Key` in my substrate implementation

File: src/keys.rs [hunk 4]
```diff
@@ -65,10 +67,12 @@ impl<PRFH: Hasher, MSGH: Hasher> Key<PRFH, MSGH> {
...
-        let p_seed = self.p_seed;
-        let (public_key, _) =
-            self.params
-                .compute_ladders(&p_seed, None, &self.secret_key, false, false)?;
+        let (public_key, _) = self.params.compute_ladders(
+            &self.p_seed,
+            None,
+            &self.secret_key,
+            ComputeLaddersMode::ComputePublicKey,
+        )?;
```
Comment (clearloop) [inline src/keys.rs:75]: if it is possible compute the public key in `new` or `generate`, so we don't need to use `Option` for the public key anymore

File: src/params.rs [hunk 5]
```diff
@@ -114,21 +135,20 @@ impl<PRFH: Hasher, MSGH: Hasher> Params<PRFH, MSGH> {
...
         hasher.sum(&mut msg_buf);
-        hashed_msg[0..self.m as usize].clone_from_slice(&msg_buf[0..self.m as usize]);
+        hashed_msg[0..self.m].clone_from_slice(&msg_buf[0..self.m]);
         hashed_msg.append(&mut checksum(&hashed_msg));
         hashed_msg
     }
 
     pub fn compute_ladders(
-        &mut self,
+        &self,
```
Comment (clearloop) [inline src/params.rs:147]: since this is under `&self` for now, we should use `&self` for `sign` and `public_key` also

**Review by clearloop — CHANGES_REQUESTED** [review]
lgtm! 

please use `&self` for `sign` and `public_key` since `compute_ladders` is under `&self` now

File: src/params.rs (hunk 2 above)
Comment (noot) [reply src/params.rs:287]: `chains` is declared in the calling function `compute_ladders` so the size should always be ok, but I can still add it if you'd like!

File: src/params.rs (hunk 5 above)
Comment (noot) [reply src/params.rs:147]: updated

File: src/keys.rs (hunk 4 above)
Comment (noot) [reply src/keys.rs]: updated

File: src/params.rs (hunk 1 above)
Comment (noot) [reply src/params.rs]: updated to be `range: (u8, u8)`

File: src/keys.rs (hunk 3 above)
Comment (noot) [reply src/keys.rs]: yes, updated!

**Review by clearloop — APPROVED** [review]
Great work!

**Pull request merged**